Use the new error package

duglin opened the pull request:
This is the first step in converting our static strings into well-defined
error types.  This shows just a few examples of it to get a feel for how things
will look. Once we agree on the basic outline we can then work on converting
the rest of the code over.

ping @crosbymichael @stevvooe 

Signed-off-by: Doug Davis dug@us.ibm.com

Commit:
Use the new error package

This is the first step in converting out static strings into well-defined
error types.  This shows just a few examples of it to get a feel for how things
will look. Once we agree on the basic outline we can then work on converting
the rest of the code over.

Signed-off-by: Doug Davis <dug@us.ibm.com>

vdemeester commented:
@duglin this looks good :+1:

runcom commented:
I feel we're mixing api and daemon errors here. The daemon is importing the api errors where instead it should know nothing about http and controller/view layer things. Distribuition errors are awesome but they should be used in controllers whether the daemon should have an error pkg defining it's own errors - and these errors should then be mapped to distr errors (I see this logic directly in the controllers, as @cpuguy83 was doing in his api error pr,  or in an api pkg) 
This way the daemon (service) depends on api pkg (controllers) implementation. 
Tell me what do you think, my thoughts are: 1) have a daemon pkg errors defining  it's own errors types (this will be long I know) 2) have an api pkg defining it's own errors based on service/daemon errors and some mapping logic

duglin commented:
Couple of things:
- The location of where each error is defined can be moved - that's a minor thing to me. I put them in api/error because 1) it seemed like a nice common spot - easily shared between cli and daemon, and 2) eventually when the CLI imports the errors I wanted it to import as few packages as possible. Having to have a lot of imports on the CLI seemed like a real pain - while just adding one import ( api/errors ) for each api/client/*.go file seemed much more manageable.
- I think the word "api" in the import may be throwing people off, making it look like the daemon/*.go code need to know about the API layer - so perhaps I should just move api/errors to just be a top level package (e.g. just "errors").
- I really want to move the "errcode" logic out from under "distribution" because otherwise it appears as though everyone (even cli) needs to know about the registry, and clearly it doesn't.  But we haven't decided where it should go yet.

cpuguy83 commented:
My reservations with this is we aren't actually changing the situation much.
We're still reading error strings instead of having real typed errors.

duglin commented:
@cpuguy83 reading errors strings where?

duglin commented:
@cpuguy83 see: https://github.com/docker/docker/pull/15069/files#diff-44a527ea72ea9572e1cf71c7f1e2a0f0R224

stevvooe commented on api/server/server.go:
```diff
@@ -194,28 +195,47 @@ func httpError(w http.ResponseWriter, err error) {
+
+	// The first part of this if-statement will be removed once we've
+	// converted everything over to use the errcode package
+	if daError, ok := err.(errcode.Error); !ok {
```
Start using type switches for this kind of thing. It is easier to see which types of errors are broken out.

stevvooe commented:
@cpuguy83 @runcom @duglin I think we are all saying and trying the same thing here and need to sync up on the details.

In general, the `errcode` package is for defining errors at the API level, although the definitions can be used to make identity comparisons, if that fits the interface design. Wherever we define the "API" layer is probably up for discussion.

duglin commented:
@stevvooe yea we were talking about this earlier today on irc and I didn't get a chance to add more comments here, but I agree that I think we're all pretty much on the same page.

What I think we need to do is:
1 - decide if the basic layout of this usage of the errcode package is ok for the engine
2 - look to convert the engine's static string errors to ErrorCodes
3 - look to convert the engine's "pkg" packages to use well-typed errors - but this can be done as a separate work item

stevvooe commented:
@duglin My only worry here is that we let the `errcode` errors penetrate too deeply into the package hierarchy. If you look at the distribution code base, we actually define identity errors (`var ErrFoo = fmt.Errorf("")`) in each package and then map them in the API layer. This is a solid convention and decouples the API and package-level errors.

The challenge in the docker code base is defining where that API layer belongs. Arguments could be made that it only belongs in http handlers or in `daemon.go`, since certain methods are built exclusively for serving API requests.

duglin commented:
@stevvooe if I'm following you, I'm not keen on the idea of generating an error, say, in the builder that is then mapped at the API layer.  Let's say we have 1000 errors in the engine, do we really want to have 1000 mappings at the API layer?  Seems really inefficient.

I don't see the issue with the engine's components defining an ErrCode and that is bubbled all the way back up to the API layer which just returns it, and its http status code, back to the CLI.  This also allows for any parts of the engine to check for any particular error via its ErrCode instead of doing string compares- which is what we're trying to avoid.

stevvooe commented:
@duglin There aren't 1000 mappings. There a several methods that expect certain package-level errors which get checked and converted to API errors in the API code. The idea here is to maintain proper package separation. The build layer should have no clue about API errors.

duglin commented:
Let's take a very specific example that I just added to the PR to make sure we're talking about the same thing.
See:  https://github.com/docker/docker/pull/15069/files#diff-9ead3fedd50f7ee2684c132d647703b6L44
are you saying that you would prefer that it continue to use fmt.Errorf() in the builder there instead of an ErrCode ?

stevvooe commented:
@duglin That example falls into the "since certain methods are built exclusively for serving API requests." This is an open question. Since `daemon.go` is positioned to answer API requests, these should probably be `errcode` errors.

Typically, when I talk about identity errors, I am referring to this:

```go
var ErrMyError = fmt.Errorf("a condition")
```

These can be checked with:

```go
if err == ErrMyError { // ... handle it }
```

Type-based errors are also acceptable if more error information is required. These use cases are package-level decisions and aren't under the purview of the `errcode` package, which is focused on API.

Small nit: all error codes should have the convention of being prefixed with `ErrorCode`. For example, these should be `ErrorCodeNoSuchContainer`.

duglin commented:
@stevvooe I actually think we may be agreeing. Looking at https://github.com/docker/docker/pull/15069/files#diff-1a1f3e7ad9b1d7584e2d3e7d0c4c3db9R129 I think its an example of the other case you're referring to.  In this case I think the truncindex package should define a typed error (not errcode) and then daemon.go maps it to an errCode.  Sound consistent with what you're thinking?

stevvooe commented:
@duglin Indeed.

We should probably take a crack at trying to explain this concept in the package doc.

duglin commented:
@stevvooe I added a README to it and added "ErrorCode" to the errors.

ping @crosbymichael are you ok with this general direction? If so we can start the process of converting the static strings.

ankushagarwal commented on api/errors/README.md:
```diff
@@ -0,0 +1,58 @@
+will be used to populate the HTTP status code. If not present the default
+value will be `StatusInternalServerError`, 500.
+
+Not all errors generated within the engine's executable will be propigated
```
propigated - propagated

duglin replied:
@ankushagarwal thanks - fixed!

stevvooe commented:
@duglin Add that to the package docs for `errcode`. No one will ever find a random readme. It will show up [here](http://godoc.org/github.com/docker/distribution/registry/api/errcode) if done correctly.

duglin commented:
Yup it's on my todo list. Just been a busy day. 

-Doug

Sent from my iPhone

> On Jul 30, 2015, at 5:24 PM, Stephen Day <notifications@github.com> wrote:
> 
> @duglin Add that to the package docs for errcode. No one will ever find a random readme. It will show up here if done correctly.
> 
> —
> Reply to this email directly or view it on GitHub.
>

duglin commented:
This can not be merged until we get a new errcode package to fix some stuff.

stevvooe commented on api/errors/builder.go:
```diff
@@ -0,0 +1,93 @@
+var (
+	// ErrorCodeAtLeastOneArg is generated when the parser comes across a
+	// Dockerfile command that doesn't have any args.
+	ErrorCodeAtLeastOneArg = errcode.Register(errGroup, errcode.ErrorDescriptor{
```
This seems like a mis-use of the `errcode` package. Are there different actions a client might take under each case? Seems like there are only a few real conditions:

1. `ArgumentError`
2. `SyntaxError`
3. `Unsupported`

duglin replied:
I'm not following, how is this is mis-use?  All I did was convert existing errors into ErrCodes. While we may be able to tweak the error messages, I don't think that should be the point of this initial exercise. I'd prefer to have zero UX impact on the first pass of this, meaning just get errcode in there.  Then we can revisit, and potentially consolidate, error messages, but that's just a clean-up/optimization to me.  Or are you referring to something else?

stevvooe replied:
I mean that it makes more sense to define a few classes of errors and divide them with solid messaging, than to make a code for every condition. Effectively, it is easier to add them than to taken them away.

In this case, we really want to have an error code not for each condition, but rather, for each action a consumer of these functions can take action upon. We need to ask the question, can a client or callee take different action based receiving a `EXACTLYONEARG` or `ATLEASTONEARG`? If not, having different codes for these conditions only serves to make the error handling more complex.

duglin replied:
One of the purposes of this is to have a static string for each error message so that they can, eventually, be translated.  You seem to be pushing for one error code mapping to potentially many error strings. That will break i18n where they need to have an exact 1-1 mapping between error code/ID and a string.  Additionally, with a 1-M mapping, I think you mean to have the exact error string still be hard-coded into the code itself, meaning its (sort of) dynamic and not static. Again, that breaks the ability to programmatically know exactly what error we have.

With i18n an end-user needs to be able to see an error message in some language that they can't read, but take the error ID and look it up in some docs and know exactly what that message says in their chosen language.  They can't do that if its a 1-M mapping.

stevvooe replied:
The primary goal of the API package is consistent error reporting and json error format. I18N is really a UI issue. We should be considering this use case, but not overly optimizing the use of `errcode` for i18n.

This is why we allowed different messages in the original `errcode`. Otherwise, the mapping of codes and actions becomes influenced by the UI rather than sane error handling.

duglin replied:
Mapping of actions from error code doesn't seem in scope for this as we can't possibly know what actions people should take - that's a UI, or CLI type of consideration.  All we can do is present the caller with the unique error for each condition - they can then decide what to do.  If we allow two different conditions to share the same error then there's ambiguity and that's not good.

duglin replied:
Using your example of EXACTLYONEARG vs ATLEASTONEARG - yes I can see taking very different actions based on this information.  One might present the user with one entry field, while the 2nd may present them with a list of entry fields.  But the point is that we shouldn't preclude receivers of these errors from making that choice themselves by merging error codes for different error situations.

duglin commented:
Updated this PR based on the current errcode stuff from distribution already in the engine's repo. If/when the changes in https://github.com/docker/distribution/pull/799 are made then I'll twiddle things but I don't want to wait any longer for the core notion of this PR to be reviewed and merged.

ping @crosbymichael @icecrime please review the basic design to see if you're ok with the direction it is headed.

thaJeztah commented:
ping @icecrime @crosbymichael ^^

duglin commented:
ping @crosbymichael @icecrime

stevvooe commented:
@duglin I can't see docker/distribution#799 getting merged. This PR is a step in the right direction, but we need to do some massaging.

duglin commented:
@stevvooe need more info on what needs to me massaged.

stevvooe commented on api/errors/error.go:
```diff
@@ -0,0 +1,25 @@
+	// is called and we just need to bubble the unknown error back up to
+	// the API. We should try to avoid using this.
+	ErrorCodeGeneric = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "GENERIC",
```
This is the role of the `UNKNOWN` error code.

Besides, remember, there are no generics in Go. :stuck_out_tongue_winking_eye:

duglin replied:
updated.

duglin commented:
Reviewers: Just to repeat what I said in the first comment.... This is the first step in converting out static strings into well-defined error types. It has the framework that we'll use and converts a fair number of the builder error messages over to this structure so you can see what changed from a docker dev POV.

If approved, the rest of the static strings can be converted over and so it should be a list of PR that have things like you'll see here (for using the lookup error):
https://github.com/docker/docker/pull/15069/files#diff-9ead3fedd50f7ee2684c132d647703b6L44

and here (for defining a new error string):
https://github.com/docker/docker/pull/15069/files#diff-7bfc5127b9f40911fd580c147ed2faf7R69

icecrime commented:
I love the idea, this is much cleaner.

However... I have to say I agree with https://github.com/docker/docker/pull/15069/files#r37594457. Converting error 1-to-1 is not ideal, especially when a class of error is very obvious (e.g., wrong number of arguments). I also don't believe i18n should have any weight in this conversation: any reasonably smart client/UX could generate a localized message given a typed error alongside an error specific argument.

So my 2cts:
- Yes! Go for it :+1:
- Let's try and rationalize the error classes rather than make a 1-1 mapping

duglin commented:
@icecrime Just to be clear on the initial work:
- each existing unique error string in the engine has its own unique error code/object. This is the 1-1 mapping. However, some error strings are used in multiple places, so that's a 1-M mapping (string->usage) and that mapping didn't change in this string->errorCode conversion.

- as I mentioned previously, if we want to consolidate errors such that the same errorCode/string is used in more places (thus reducing the number of unique errorCode/strings) that's fine but I'd prefer to do that after the initial work is completed.  My goal with the first pass of this work is to have **zero** UX impact and simply get the infrastructure in place to stop using static strings in the code.

bfirsh commented:
ping @icecrime - needs some confirmation from you :)

duglin commented:
ping @crosbymichael @icecrime

duglin commented:
ping @icecrime @crosbymichael

icecrime commented:
Fine by me regarding the previous comment, I just want to reinforce that we should consolidate errors in the future ;-)

icecrime commented:
LGTM

jfrazelle commented:
cool LGTM

jfrazelle commented:
ping @calavera @vieux @aluzzardi whoever for api things :)

vieux commented on api/server/server.go:
```diff
@@ -187,28 +188,56 @@ func httpError(w http.ResponseWriter, err error) {
+	errMsg := err.Error()
+
+	switch err.(type) {
+	case errcode.ErrorCode:
```
Could you please explain here on add a comment on why we might receive `errcode.ErrorCode` or `errcode.Error` I don't really get it.

duglin replied:
@vieux I added a comment to the code so that it'll live on beyond this PR - see if that explains it well enough. If not I can add more.

vieux commented:
besides my comment, LGTM

stevvooe commented on api/errors/error.go:
```diff
@@ -0,0 +1,25 @@
+	// available. This will typically happen when some low-level function
+	// is called and we just need to bubble the unknown error back up to
+	// the API. We should try to avoid using this.
+	ErrorCodeGeneric = errcode.Register(errGroup, errcode.ErrorDescriptor{
```
Please use `errcode.ErrorCodeUnknown` for this purpose.

duglin replied:
The error message associated with that error code isn't appropriate for this engine situation - I wanted something more specific to the engine - and to say something about the logs, which the errorcode package probably can't assume in general.

stevvooe replied:
Then let's make it more specific. I really don't like have a bunch of the same things defined because you slightly disagree. We have an error code defined globally for this exact use case. Use it.

duglin replied:
The error message from the generic errorcode isn't really that helpful in this case so I made one that's better suited for the engine.  To talk about "API classifications" won't really mean much to the end-user and, in fact, could make it sound like there's an error in Docker because it didn't "classify" all of its possible errors.  Granted, its subjective. A generic package like errorcode can't assume the text it picked for its canned errors are good enough for everyone in all cases, and in this case its not.  I can rename it if you want so it doesn't look like we're duplicating the errocode one, just having a hard time thinking of a choice.

stevvooe replied:
I see. `ErrorCodeEngineUnknown`?

I think part of the problem is that this use case is tying error codes too tightly to the UI. Frankly, I don't see what the value of redefining this from "unknown error" with a useful message to "An unexpected error occurred" (ignoring the fact that this error message violates [Go's style guide](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)). The way the error code was structured, errors of this nature would appear as "unknown error: something unfortunate happened". Depending on how careful the error chaining is managed, usually the problem can be succinctly identified.

duglin replied:
I decided to just remove this error definition for now. As I was changing to your suggested name I realized that I don't actually use it yet - I just added it with the assumption that I would need it as some point. But, let's wait until we need such a generic error message and worry about it then.

duglin commented:
janky is happy - think we're good to go
ping @jfrazelle for the merge

cpuguy83 commented on api/errors/daemon.go:
```diff
@@ -0,0 +1,21 @@
+	// name or ID and we can't find it.
+	ErrorCodeNoSuchContainer = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "NOSUCHCONTAINER",
+		Message:        "no such id: %s",
```
shouldn't this be `no such id or name: %s`, or `no such container: %s`?

Also wondering if this could be grouped with other types since we have lookups for containers, images, volumes, and exec instances all with extremely similar cases.
Let me know if that's not the right usage of the package.

duglin replied:
@cpuguy83 yes we can make those types of changes but I'd like to do that after the initial pass of just moving the error messages we have today into this package/format. We can get into a lot of bike shedding while trying to twiddle error messages and consolidate them, and so I'd like to do that in subsequent PRs. My main goal of this PR is to have zero UX impact and keep all changes internal - and to not break anything/anyone.

cpuguy83 commented on api/server/server.go:
```diff
@@ -187,28 +188,71 @@ func httpError(w http.ResponseWriter, err error) {
+	// status code associated with it, and if so use it.
+	switch err.(type) {
+	case errcode.ErrorCode:
+		daError, _ := err.(errcode.ErrorCode)
```
These variable names are super weird to me, can you explain the `da` part?
Mostly I read this like Jar Jar Binks in my head.

duglin replied:
LOL "da" as in the "the" with a bit of an accent.  :-)
that's just my way of saying "the error" w/o using a variable called "error" or "err" since those words are already taken.  :-)    I can switch it if you want.

thaJeztah replied:
`thaError`? :wink:

duglin replied:
LOL

cpuguy83 commented:
One legit question, one nit.

cpuguy83 commented:
LGTM
